Exception when building cachemanager in parallel

Client version:
```ehcache-clustered
Implementation-Version: 3.1.0-SNAPSHOT 3d44aa4c4a3047b773beab6ee761262
 036491394
```

Server version:
```
Terracotta 5.0.0-beta2, as of 2016-06-30 at 01:31:45 PDT (Revision 94748627d9b6fb3e06e6ccad724f821ef1ef741d from UNKNOWN)
```

Configuration:
````java
 final CacheManagerBuilder<PersistentCacheManager> clusteredCacheManagerBuilder = CacheManagerBuilder.newCacheManagerBuilder()
        .with(ClusteringServiceConfigurationBuilder.cluster(CLUSTER_URI).autoCreate()
        .defaultServerResource("primary-server-resource")
        .resourcePool("resource-pool-a", 64, MemoryUnit.MB)
        .resourcePool("resource-pool-b", 64, MemoryUnit.MB, "secondary-server-resource"))
        .withCache(CACHE_NAME, CacheConfigurationBuilder.newCacheConfigurationBuilder(Long.class, AddressProto.Address.class,
            ResourcePoolsBuilder.newResourcePoolsBuilder()
                .with(ClusteredResourcePoolBuilder.fixed("primary-server-resource", 64, MemoryUnit.MB)))
            .add(new ClusteredStoreConfiguration(Consistency.STRONG)));

final PersistentCacheManager cacheManager = clusteredCacheManagerBuilder.build(true);
````

When ran the clients in parallel, hits the exception. Complete trace as below,

```
   [testng] 20:44:06.355 [main] DEBUG org.ehcache.core.EhcacheManager - Creating Cache 'clustered-cache' in EhcacheManager.
   [testng] 20:44:06.370 [main] DEBUG o.e.i.i.s.s.DefaultSerializationProvider - Serializer for <java.lang.Long> : org.ehcache.impl.serialization.LongSerializer@1c1cfa64
   [testng] 20:44:06.370 [main] DEBUG o.e.c.i.service.ServiceLocator - All Services successfully started.
   [testng] 20:44:06.402 [main] ERROR o.e.c.c.i.s.DefaultClusteringService - Classpath issue - missing entity provider
   [testng] 20:44:06.370 [main] DEBUG org.ehcache.core.EhcacheManager - Creating Cache 'clustered-cache' in EhcacheManager.
   [testng] org.terracotta.exception.EntityAlreadyExistsException: Entity: org.terracotta.entity.map.common.ConcurrentClusteredMap:clustered-cache-DefaultSerializationProvider-Value-CompactJavaSerializer-ObjectStreamClassIndex already exists
   [testng] 20:44:06.370 [main] DEBUG o.e.i.i.s.s.DefaultSerializationProvider - Serializer for <java.lang.Long> : org.ehcache.impl.serialization.LongSerializer@4ce67c1e
   [testng]     at com.terracotta.connection.entity.TerracottaEntityRef.create(TerracottaEntityRef.java:152) ~[ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng] 20:44:06.418 [main] DEBUG o.e.i.i.s.s.DefaultSerializationProvider - Serializer for <com.terracotta.qa.ehcache.generator.AddressProto$Address> : org.ehcache.impl.serialization.CompactJavaSerializer@1a4c8517
   [testng]     at org.ehcache.clustered.client.internal.service.DefaultClusteringService.createConcurrentClusteredMap(DefaultClusteringService.java:378) [ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.clustered.client.internal.service.DefaultClusteringService.getConcurrentMap(DefaultClusteringService.java:364) [ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.clustered.client.internal.service.ClusteredStateRepository.getPersistentConcurrentMap(ClusteredStateRepository.java:42) [ehcache-clustered-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.impl.serialization.CompactJavaSerializer.<init>(CompactJavaSerializer.java:80) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) [na:1.7.0_79]
   [testng]     at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57) [na:1.7.0_79]
   [testng]     at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) [na:1.7.0_79]
   [testng]     at java.lang.reflect.Constructor.newInstance(Constructor.java:526) [na:1.7.0_79]
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$AbstractProvider.constructSerializer(DefaultSerializationProvider.java:281) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$PersistentProvider.createSerializer(DefaultSerializationProvider.java:205) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$AbstractProvider.createValueSerializer(DefaultSerializationProvider.java:254) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider.createValueSerializer(DefaultSerializationProvider.java:107) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.core.EhcacheManager.getStore(EhcacheManager.java:462) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.core.EhcacheManager.createNewEhcache(EhcacheManager.java:308) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:257) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:564) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at org.ehcache.config.builders.CacheManagerBuilder.build(CacheManagerBuilder.java:72) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
   [testng]     at com.terracotta.qa.ehcache.test.MultipleClientsTest.testCRUDOps(MultipleClientsTest.java:96) [test-classes/:na]
   [testng]     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_79]
   [testng]     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_79]
   [testng]     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_79]
   [testng]     at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_79]
   [testng]     at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85) [testng-6.8.17.jar:na]
   [testng]     at org.testng.internal.MethodInvocationHelper$1.runTestMethod(MethodInvocationHelper.java:200) [testng-6.8.17.jar:na]
   [testng]     at org.jboss.byteman.contrib.bmunit.BMNGAbstractRunner.run(BMNGAbstractRunner.java:58) [byteman-bmunit-2.1.2.jar:na]
   [testng]     at org.testng.internal.MethodInvocationHelper.invokeHookable(MethodInvocationHelper.java:212) [testng-6.8.17.jar:na]
   [testng]     at org.testng.internal.Invoker.invokeMethod(Invoker.java:689) [testng-6.8.17.jar:na]
   [testng]     at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882) [testng-6.8.17.jar:na]
   [testng]     at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189) [testng-6.8.17.jar:na]
   [testng]     at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124) [testng-6.8.17.jar:na]
   [testng]     at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestRunner.privateRun(TestRunner.java:767) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestRunner.run(TestRunner.java:617) [testng-6.8.17.jar:na]
   [testng]     at org.testng.SuiteRunner.runTest(SuiteRunner.java:348) [testng-6.8.17.jar:na]
   [testng]     at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343) [testng-6.8.17.jar:na]
   [testng]     at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305) [testng-6.8.17.jar:na]
   [testng]     at org.testng.SuiteRunner.run(SuiteRunner.java:254) [testng-6.8.17.jar:na]
   [testng]     at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52) [testng-6.8.17.jar:na]
   [testng]     at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestNG.runSuitesLocally(TestNG.java:1149) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestNG.run(TestNG.java:1057) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestNG.privateMain(TestNG.java:1364) [testng-6.8.17.jar:na]
   [testng]     at org.testng.TestNG.main(TestNG.java:1333) [testng-6.8.17.jar:na]
   [testng] 20:44:06.482 [main] INFO  o.e.c.c.i.s.DefaultClusteringService - stop called for clustered tiers on terracotta://localhost:9510
   [testng] 20:44:06.482 [main] ERROR org.ehcache.core.EhcacheManager - Initialize failed.
   [testng] 20:44:06.560 [main] INFO  org.ehcache.core.EhcacheManager - Cache 'clustered-cache' created in EhcacheManager.
   [testng] org.ehcache.StateTransitionException: Cache 'clustered-cache' creation in EhcacheManager failed.
   [testng] 20:44:06.560 [main] DEBUG org.ehcache.core.EhcacheManager - Initialize successful.
   [testng]     at org.ehcache.core.StatusTransitioner$Transition.failed(StatusTransitioner.java:235)
   [testng] 20:44:06.560 [main] INFO  c.t.q.e.test.MultipleClientsTest - Going to await for other threads before put
   [testng]     at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:585)
   [testng]     at org.ehcache.config.builders.CacheManagerBuilder.build(CacheManagerBuilder.java:72)
   [testng]     at com.terracotta.qa.ehcache.test.MultipleClientsTest.testCRUDOps(MultipleClientsTest.java:96)
   [testng]     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   [testng]     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
   [testng]     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   [testng]     at java.lang.reflect.Method.invoke(Method.java:606)
   [testng]     at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
   [testng]     at org.testng.internal.MethodInvocationHelper$1.runTestMethod(MethodInvocationHelper.java:200)
   [testng]     at org.jboss.byteman.contrib.bmunit.BMNGAbstractRunner.run(BMNGAbstractRunner.java:58)
   [testng]     at org.testng.internal.MethodInvocationHelper.invokeHookable(MethodInvocationHelper.java:212)
   [testng]     at org.testng.internal.Invoker.invokeMethod(Invoker.java:689)
   [testng]     at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
   [testng]     at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
   [testng]     at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
   [testng]     at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
   [testng]     at org.testng.TestRunner.privateRun(TestRunner.java:767)
   [testng]     at org.testng.TestRunner.run(TestRunner.java:617)
   [testng]     at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
   [testng]     at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
   [testng]     at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
   [testng]     at org.testng.SuiteRunner.run(SuiteRunner.java:254)
   [testng]     at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
   [testng]     at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
   [testng]     at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
   [testng]     at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
   [testng]     at org.testng.TestNG.run(TestNG.java:1057)
   [testng]     at org.testng.TestNG.privateMain(TestNG.java:1364)
   [testng]     at org.testng.TestNG.main(TestNG.java:1333)
   [testng] Caused by: java.lang.IllegalStateException: Cache 'clustered-cache' creation in EhcacheManager failed.
   [testng]     at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:285)
   [testng]     at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:564)
   [testng]     ... 28 more
   [testng] Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$AbstractProvider.constructSerializer(DefaultSerializationProvider.java:291)
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$PersistentProvider.createSerializer(DefaultSerializationProvider.java:205)
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$AbstractProvider.createValueSerializer(DefaultSerializationProvider.java:254)
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider.createValueSerializer(DefaultSerializationProvider.java:107)
   [testng]     at org.ehcache.core.EhcacheManager.getStore(EhcacheManager.java:462)
   [testng]     at org.ehcache.core.EhcacheManager.createNewEhcache(EhcacheManager.java:308)
   [testng]     at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:257)
   [testng]     ... 29 more
   [testng] Caused by: java.lang.reflect.InvocationTargetException
   [testng]     at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
   [testng]     at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
   [testng]     at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
   [testng]     at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
   [testng]     at org.ehcache.impl.internal.spi.serialization.DefaultSerializationProvider$AbstractProvider.constructSerializer(DefaultSerializationProvider.java:281)
   [testng]     ... 35 more
   [testng] Caused by: java.lang.AssertionError: Classpath issues as expected entity is not resolvable
   [testng]     at org.ehcache.clustered.client.internal.service.DefaultClusteringService.createConcurrentClusteredMap(DefaultClusteringService.java:387)
   [testng]     at org.ehcache.clustered.client.internal.service.DefaultClusteringService.getConcurrentMap(DefaultClusteringService.java:364)
   [testng]     at org.ehcache.clustered.client.internal.service.ClusteredStateRepository.getPersistentConcurrentMap(ClusteredStateRepository.java:42)
   [testng]     at org.ehcache.impl.serialization.CompactJavaSerializer.<init>(CompactJavaSerializer.java:80)
   [testng]     ... 40 more
````

Tested the on Ehcache v3.1.1 and I'm not getting the stack trace that is reported here. Instead I'm getting this:

```
[pool-1-thread-2] ERROR org.ehcache.core.EhcacheManager - Initialize failed.
org.ehcache.StateTransitionException: Cache 'foo' creation in EhcacheManager failed.
	at org.ehcache.core.StatusTransitioner$Transition.failed(StatusTransitioner.java:235)
	at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:585)
	at org.ehcache.config.builders.CacheManagerBuilder.build(CacheManagerBuilder.java:72)
	at org.ehcache.Test$1.run(Test.java:45)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: Cache 'foo' creation in EhcacheManager failed.
	at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:285)
	at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:564)
	... 7 more
Caused by: org.ehcache.StateTransitionException: Unable to create clustered tier proxy - DefaultClusterCacheIdentifier@foo
	at org.ehcache.core.StatusTransitioner$Transition.succeeded(StatusTransitioner.java:209)
	at org.ehcache.core.Ehcache.init(Ehcache.java:582)
	at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:258)
	... 8 more
Caused by: java.lang.RuntimeException: Unable to create clustered tier proxy - DefaultClusterCacheIdentifier@foo
	at org.ehcache.clustered.client.internal.store.ClusteredStore$Provider.initStore(ClusteredStore.java:566)
	at org.ehcache.core.EhcacheManager$8.init(EhcacheManager.java:496)
	at org.ehcache.core.StatusTransitioner.runInitHooks(StatusTransitioner.java:135)
	at org.ehcache.core.StatusTransitioner.access$000(StatusTransitioner.java:33)
	at org.ehcache.core.StatusTransitioner$Transition.succeeded(StatusTransitioner.java:194)
	... 10 more
Caused by: org.ehcache.CachePersistenceException: Unable to create clustered tier proxy 'foo' for entity 'my-application'
	at org.ehcache.clustered.client.internal.service.DefaultClusteringService.getServerStoreProxy(DefaultClusteringService.java:383)
	at org.ehcache.clustered.client.internal.store.ClusteredStore$Provider.initStore(ClusteredStore.java:564)
	... 14 more
Caused by: org.ehcache.clustered.client.internal.service.ClusteredTierCreationException: Error creating clustered tier 'foo'
	at org.ehcache.clustered.client.internal.EhcacheClientEntity.createCache(EhcacheClientEntity.java:181)
	at org.ehcache.clustered.client.internal.service.DefaultClusteringService.getServerStoreProxy(DefaultClusteringService.java:369)
	... 15 more
Caused by: org.ehcache.clustered.common.internal.exceptions.InvalidStoreException: Clustered tier 'foo' already exists
	at org.ehcache.clustered.common.internal.exceptions.InvalidStoreException.withClientStackTrace(InvalidStoreException.java:39)
	at org.ehcache.clustered.common.internal.exceptions.InvalidStoreException.withClientStackTrace(InvalidStoreException.java:22)
	at org.ehcache.clustered.common.internal.messages.ResponseCodec.decode(ResponseCodec.java:117)
	at org.ehcache.clustered.common.internal.messages.EhcacheCodec.decodeResponse(EhcacheCodec.java:67)
	at org.ehcache.clustered.common.internal.messages.EhcacheCodec.decodeResponse(EhcacheCodec.java:22)
	at com.tc.object.EntityClientEndpointImpl$InvocationBuilderImpl$1.getWithTimeout(EntityClientEndpointImpl.java:174)
	at com.tc.object.EntityClientEndpointImpl$InvocationBuilderImpl$1.getWithTimeout(EntityClientEndpointImpl.java:156)
	at org.ehcache.clustered.client.internal.EhcacheClientEntity.waitFor(EhcacheClientEntity.java:276)
	at org.ehcache.clustered.client.internal.EhcacheClientEntity.invokeInternal(EhcacheClientEntity.java:240)
	at org.ehcache.clustered.client.internal.EhcacheClientEntity.createCache(EhcacheClientEntity.java:179)
	... 16 more
Caused by: org.ehcache.clustered.common.internal.exceptions.InvalidStoreException: Clustered tier 'foo' already exists
	at org.ehcache.clustered.server.ServerStoreManager.createStore(ServerStoreManager.java:277)
	at org.ehcache.clustered.server.EhcacheActiveEntity.createServerStore(EhcacheActiveEntity.java:587)
	at org.ehcache.clustered.server.EhcacheActiveEntity.invokeLifeCycleOperation(EhcacheActiveEntity.java:312)
	at org.ehcache.clustered.server.EhcacheActiveEntity.invoke(EhcacheActiveEntity.java:269)
	at org.ehcache.clustered.server.EhcacheActiveEntity.invoke(EhcacheActiveEntity.java:75)
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$performAction$38(ManagedEntityImpl.java:607)
	at com.tc.objectserver.entity.ManagedEntityImpl$$Lambda$19/571772296.run(Unknown Source)
	at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:259)
	at com.tc.objectserver.entity.ManagedEntityImpl.performAction(ManagedEntityImpl.java:607)
	at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:355)
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$processInvokeRequest$31(ManagedEntityImpl.java:204)
	at com.tc.objectserver.entity.ManagedEntityImpl$$Lambda$18/1276700556.run(Unknown Source)
	at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.run(ManagedEntityImpl.java:834)
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.invoke(RequestProcessor.java:216)
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.run(RequestProcessor.java:208)
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:27)
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:23)
	at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502)
	at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)
```

Server version:
````
Terracotta 5.0.6-beta2, as of 2016-09-02 at 17:28:46 EDT (Revision 6e7d0d902c7a04c7ef5572cf15bbc44ff7de9bb9 from UNKNOWN)
````

Client version:
````
Implementation-Version: 3.2.0-SNAPSHOT f0a3d6fcf3c82c7e2df0771ab5ff89a
 fbd5584b9
````

One of the client fails at "clusteredCacheManagerBuilder.build(true)" with the below exception (similar to the exception posted by Albin in the previous comment)


````
    barrier found by jvmid 16944@MCKEWA01 = org.terracotta.entity.barrier.TerracottaBarrier@357c4ab7
    barrier found by jvmid 18152@MCKEWA01 = org.terracotta.entity.barrier.TerracottaBarrier@719ebf
    cache =org.ehcache.core.Ehcache@5516e01c
    org.ehcache.StateTransitionException: Cache 'clustered-cache' creation in EhcacheManager failed.
        at org.ehcache.core.StatusTransitioner$Transition.failed(StatusTransitioner.java:235)
        at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:585)
        at org.ehcache.config.builders.CacheManagerBuilder.build(CacheManagerBuilder.java:72)
        at com.terracotta.qa.ehcache.test.MultipleClientsTest.initCacheManager(MultipleClientsTest.java:306)
        at com.terracotta.qa.ehcache.test.MultipleClientsTest.testCRUDOps(MultipleClientsTest.java:84)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
        at org.testng.internal.MethodInvocationHelper$1.runTestMethod(MethodInvocationHelper.java:200)
        at org.jboss.byteman.contrib.bmunit.BMNGAbstractRunner.run(BMNGAbstractRunner.java:58)
        at org.testng.internal.MethodInvocationHelper.invokeHookable(MethodInvocationHelper.java:212)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:689)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
        at org.testng.TestRunner.privateRun(TestRunner.java:767)
        at org.testng.TestRunner.run(TestRunner.java:617)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
        at org.testng.SuiteRunner.run(SuiteRunner.java:254)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
        at org.testng.TestNG.run(TestNG.java:1057)
        at org.testng.TestNG.privateMain(TestNG.java:1364)
        at org.testng.TestNG.main(TestNG.java:1333)
    Caused by: java.lang.IllegalStateException: Cache 'clustered-cache' creation in EhcacheManager failed.
        at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:285)
        at org.ehcache.core.EhcacheManager.init(EhcacheManager.java:564)
        ... 29 more
    Caused by: org.ehcache.StateTransitionException: Unable to create clustered tier proxy - DefaultClusterCacheIdentifier@clustered-cache
        at org.ehcache.core.StatusTransitioner$Transition.succeeded(StatusTransitioner.java:209)
        at org.ehcache.core.Ehcache.init(Ehcache.java:582)
        at org.ehcache.core.EhcacheManager.createCache(EhcacheManager.java:258)
        ... 30 more
    Caused by: java.lang.RuntimeException: Unable to create clustered tier proxy - DefaultClusterCacheIdentifier@clustered-cache
        at org.ehcache.clustered.client.internal.store.ClusteredStore$Provider.initStore(ClusteredStore.java:580)
        at org.ehcache.core.EhcacheManager$8.init(EhcacheManager.java:496)
        at org.ehcache.core.StatusTransitioner.runInitHooks(StatusTransitioner.java:135)
        at org.ehcache.core.StatusTransitioner.access$000(StatusTransitioner.java:33)
        at org.ehcache.core.StatusTransitioner$Transition.succeeded(StatusTransitioner.java:194)
        ... 32 more
    Caused by: org.ehcache.CachePersistenceException: Unable to create clustered tier proxy 'clustered-cache' for entity 'cachemanager'
        at org.ehcache.clustered.client.internal.service.DefaultClusteringService.getServerStoreProxy(DefaultClusteringService.java:411)
        at org.ehcache.clustered.client.internal.store.ClusteredStore$Provider.initStore(ClusteredStore.java:578)
        ... 36 more
    Caused by: org.ehcache.clustered.client.internal.service.ClusteredTierCreationException: Error creating clustered tier 'clustered-cache'
        at org.ehcache.clustered.client.internal.EhcacheClientEntity.createCache(EhcacheClientEntity.java:189)
        at org.ehcache.clustered.client.internal.service.DefaultClusteringService.getServerStoreProxy(DefaultClusteringService.java:397)
        ... 37 more
    Caused by: org.ehcache.clustered.common.internal.exceptions.InvalidStoreException: Clustered tier 'clustered-cache' already exists
        at org.ehcache.clustered.common.internal.exceptions.InvalidStoreException.withClientStackTrace(InvalidStoreException.java:39)
        at org.ehcache.clustered.common.internal.exceptions.InvalidStoreException.withClientStackTrace(InvalidStoreException.java:22)
        at org.ehcache.clustered.common.internal.messages.ResponseCodec.decode(ResponseCodec.java:126)
        at org.ehcache.clustered.common.internal.messages.EhcacheCodec.decodeResponse(EhcacheCodec.java:83)
        at org.ehcache.clustered.common.internal.messages.EhcacheCodec.decodeResponse(EhcacheCodec.java:26)
        at com.tc.object.EntityClientEndpointImpl$InvocationBuilderImpl$1.getWithTimeout(EntityClientEndpointImpl.java:174)
        at com.tc.object.EntityClientEndpointImpl$InvocationBuilderImpl$1.getWithTimeout(EntityClientEndpointImpl.java:156)
        at org.ehcache.clustered.client.internal.EhcacheClientEntity.waitFor(EhcacheClientEntity.java:302)
        at org.ehcache.clustered.client.internal.EhcacheClientEntity.invokeInternal(EhcacheClientEntity.java:265)
        at org.ehcache.clustered.client.internal.EhcacheClientEntity.createCache(EhcacheClientEntity.java:186)
        ... 38 more
    Caused by: org.ehcache.clustered.common.internal.exceptions.InvalidStoreException: Clustered tier 'clustered-cache' already exists
        at org.ehcache.clustered.server.EhcacheStateServiceImpl.createStore(EhcacheStateServiceImpl.java:292)
        at org.ehcache.clustered.server.EhcacheActiveEntity.createServerStore(EhcacheActiveEntity.java:586)
        at org.ehcache.clustered.server.EhcacheActiveEntity.invokeLifeCycleOperation(EhcacheActiveEntity.java:326)
        at org.ehcache.clustered.server.EhcacheActiveEntity.invoke(EhcacheActiveEntity.java:237)
        at org.ehcache.clustered.server.EhcacheActiveEntity.invoke(EhcacheActiveEntity.java:79)
        at com.tc.objectserver.entity.ManagedEntityImpl.lambda$performAction$55(ManagedEntityImpl.java:621)
        at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:284)
        at com.tc.objectserver.entity.ManagedEntityImpl.performAction(ManagedEntityImpl.java:621)
        at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:378)
        at com.tc.objectserver.entity.ManagedEntityImpl.lambda$processInvokeRequest$50(ManagedEntityImpl.java:236)
        at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.run(ManagedEntityImpl.java:897)
        at com.tc.objectserver.entity.RequestProcessor$EntityRequest.invoke(RequestProcessor.java:162)
        at com.tc.objectserver.entity.RequestProcessor$EntityRequest.run(RequestProcessor.java:154)
        at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:27)
        at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:23)
        at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502)
        at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)
    Exception by 18152@MCKEWA01Client=0
````

--- a/clustered/client/src/main/java/org/ehcache/clustered/client/internal/service/DefaultClusteringService.java
+++ b/clustered/client/src/main/java/org/ehcache/clustered/client/internal/service/DefaultClusteringService.java
@@ -26,7 +26,6 @@ import org.ehcache.clustered.client.internal.EhcacheEntityCreationException;
 import org.ehcache.clustered.client.internal.EhcacheEntityNotFoundException;
 import org.ehcache.clustered.client.internal.EhcacheEntityValidationException;
 import org.ehcache.clustered.client.internal.config.ExperimentalClusteringServiceConfiguration;
-import org.ehcache.clustered.client.internal.store.ClusteredStore;
 import org.ehcache.clustered.client.internal.store.EventualServerStoreProxy;
 import org.ehcache.clustered.client.internal.store.ServerStoreProxy;
 import org.ehcache.clustered.client.internal.store.StrongServerStoreProxy;
@@ -45,7 +44,6 @@ import org.ehcache.impl.internal.concurrent.ConcurrentHashMap;
 import org.ehcache.spi.persistence.StateRepository;
 import org.ehcache.spi.service.MaintainableService;
 import org.ehcache.spi.service.Service;
-import org.ehcache.spi.service.ServiceDependencies;
 import org.ehcache.spi.service.ServiceProvider;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -382,22 +380,16 @@ class DefaultClusteringService implements ClusteringService, EntityService {
     try {
       if (configuration.isAutoCreate()) {
         try {
-          this.entity.validateCache(cacheId, clientStoreConfiguration);
-        } catch (ClusteredTierValidationException ex) {
-          if (ex.getCause() instanceof InvalidStoreException) {
-            try {
-              this.entity.createCache(cacheId, clientStoreConfiguration);
-            } catch (TimeoutException e) {
-              throw new CachePersistenceException("Unable to create clustered tier proxy '"
-                  + cacheIdentifier.getId() + "' for entity '" + entityIdentifier
-                  + "'; create operation timed out", e);
-            }
-          } else {
-            throw ex;
+          entity.createCache(cacheId, clientStoreConfiguration);
+        } catch (ClusteredTierCreationException e) {
+          // An InvalidStoreException means the cache already exists. That's fine, the validateCache will then work
+          if (!(e.getCause() instanceof InvalidStoreException)) {
+            throw e;
           }
+          entity.validateCache(cacheId, clientStoreConfiguration);
         }
       } else {
-        this.entity.validateCache(cacheId, clientStoreConfiguration);
+        entity.validateCache(cacheId, clientStoreConfiguration);
       }
     } catch (ClusteredTierException e) {
       throw new CachePersistenceException("Unable to create clustered tier proxy '" + cacheIdentifier.getId() + "' for entity '" + entityIdentifier + "'", e);

--- a/clustered/client/src/test/java/org/ehcache/clustered/client/ClusteredConcurrencyTest.java
+++ b/clustered/client/src/test/java/org/ehcache/clustered/client/ClusteredConcurrencyTest.java
@@ -1,0 +1,122 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.clustered.client;
+
+import org.ehcache.PersistentCacheManager;
+import org.ehcache.clustered.client.config.ClusteredStoreConfiguration;
+import org.ehcache.clustered.client.config.builders.ClusteredResourcePoolBuilder;
+import org.ehcache.clustered.client.config.builders.ClusteringServiceConfigurationBuilder;
+import org.ehcache.clustered.client.internal.UnitTestConnectionService;
+import org.ehcache.clustered.common.Consistency;
+import org.ehcache.config.builders.CacheConfigurationBuilder;
+import org.ehcache.config.builders.CacheManagerBuilder;
+import org.ehcache.config.builders.ResourcePoolsBuilder;
+import org.ehcache.config.units.MemoryUnit;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This test makes sure a clustered cache can be opened from many client instances. As usual with concurrency tests, a
+ * success doesn't mean it will work forever and a failure might not occur reliably. However, it puts together all
+ * conditions to make it fail in case of race condition
+ *
+ * @author Henri Tremblay
+ */
+public class ClusteredConcurrencyTest {
+
+  private static final URI CLUSTER_URI = URI.create("terracotta://example.com:9540/my-application");
+  private static final String CACHE_NAME = "clustered-cache";
+
+  private AtomicReference<Throwable> exception = new AtomicReference<Throwable>();
+
+  @Before
+  public void definePassthroughServer() throws Exception {
+    UnitTestConnectionService.add(CLUSTER_URI,
+      new UnitTestConnectionService.PassthroughServerBuilder()
+        .resource("primary-server-resource", 64, MemoryUnit.MB)
+        .resource("secondary-server-resource", 64, MemoryUnit.MB)
+        .build());
+  }
+
+  @After
+  public void removePassthroughServer() throws Exception {
+    UnitTestConnectionService.remove(CLUSTER_URI);
+  }
+
+  @Test
+  public void test() throws Throwable {
+    final int THREAD_NUM = 50;
+
+    final CountDownLatch latch = new CountDownLatch(THREAD_NUM + 1);
+
+    List<Thread> threads = new ArrayList<Thread>(THREAD_NUM);
+    for (int i = 0; i < THREAD_NUM; i++) {
+      Thread t1 = new Thread(content(latch));
+      t1.start();
+      threads.add(t1);
+    }
+
+    latch.countDown();
+    latch.await();
+
+    for(Thread t : threads) {
+      t.join();
+    }
+
+    Throwable throwable = exception.get();
+    if(throwable != null) {
+      throw throwable;
+    }
+  }
+
+  private Runnable content(final CountDownLatch latch) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        try {
+          CacheManagerBuilder<PersistentCacheManager> clusteredCacheManagerBuilder = CacheManagerBuilder.newCacheManagerBuilder()
+            .with(ClusteringServiceConfigurationBuilder.cluster(CLUSTER_URI).autoCreate()
+              .defaultServerResource("primary-server-resource")
+              .resourcePool("resource-pool-a", 32, MemoryUnit.MB)
+              .resourcePool("resource-pool-b", 32, MemoryUnit.MB, "secondary-server-resource"))
+            .withCache(CACHE_NAME, CacheConfigurationBuilder.newCacheConfigurationBuilder(Long.class, String.class,
+              ResourcePoolsBuilder.newResourcePoolsBuilder()
+                .with(ClusteredResourcePoolBuilder.clusteredDedicated("primary-server-resource", 32, MemoryUnit.MB)))
+              .add(new ClusteredStoreConfiguration(Consistency.STRONG)));
+
+          latch.countDown();
+          try {
+            latch.await();
+          } catch (InterruptedException e) {
+            // continue
+          }
+
+          clusteredCacheManagerBuilder.build(true);
+        } catch (Throwable t) {
+          exception.compareAndSet(null, t); // only keep the first exception
+        }
+      }
+    };
+  }
+}

